File: prob2.py
# Time Complexity :O(n**2)
# Space Complexity :O(n**2)
# Leet Code: Yes

class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> Optional[TreeNode]:
        if len(preorder) == 0:
            return None

        root = TreeNode(preorder[0])
        idx = inorder.index(preorder[0])
        inorderleft = inorder[:idx]
        inorderright = inorder[idx+1:]
        preorderleft = preorder[1:len(inorderleft)+1]
        preorderright = preorder[len(inorderleft)+1:]

        root.left = self.buildTree(preorderleft, inorderleft)
        root.right = self.buildTree(preorderright, inorderright)

        return root

File: prob2.java
// Time Complexity : O (n) 
// Space Complexity : O (1) no extra space
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : no


public class prob2 {

    public class TreeNode{
        int val;
        TreeNode left;
        TreeNode right;
        TreeNode() {}
        TreeNode(int val) { this.val = val; }
        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }
    int sum;

    public int sumNumbers(TreeNode root){
        if(root == null) return 0;
        helper(root, 0);
        return sum;
    }

    private void helper(TreeNode root, Integer num){
        if(root == null) return;
        helper(root.left, num * 10 + root.val);
        if(root.left == null && root.right == null){
            sum += num * 10 + root.val;
        }
        helper(root.right, num * 10 + root.val);
    }
}
